service/lib/dinstaller/dbus/storage/dasds_format_job.rb

```diff
@@ -0,0 +1,166 @@
+          attr_accessor :cylinders
+
+          # @return [Integer] number of cylinders already processed by the format operation
+          attr_accessor :progress
+
+          # @return [Boolean] whether the disk is already fully formatted
+          attr_accessor :done
+
+          # Constructor
+          #
+          # @param status [Y2S390::FormatStatus]
+          # @param dasds_tree [DasdsTree]
+          def initialize(status, dasds_tree)
+            @id = status.dasd.id
+            @cylinders = status.cylinders
+            @progress = status.progress
+            @done = status.done?
+
+            dbus_dasd = dasds_tree.find { |d| d.id == @id }
+            if dbus_dasd
+              @path = dbus_dasd.path
+            else
+              logger.warning "DASD is not longer in the D-BUS tree: #{status.inspect}"
+            end
+          end
+
+          # Progress representation as expected by the D-Bus API (property Summary and signal
+          # SummaryUpdated)
+          def to_dbus
+            [cylinders, progress, done]
+          end
+        end
+
+        JOB_INTERFACE = "org.opensuse.DInstaller.Storage1.Job"
+        private_constant :JOB_INTERFACE
+
+        dbus_interface JOB_INTERFACE do
+          dbus_reader(:running, "b")
+          dbus_reader(:exit_code, "u")
+          dbus_signal(:Finished, "exit_code:u")
+        end
+
+        DASD_FORMAT_INTERFACE = "org.opensuse.DInstaller.Storage1.DASD.Format"
+        private_constant :DASD_FORMAT_INTERFACE
+
+        dbus_interface DASD_FORMAT_INTERFACE do
+          dbus_reader(:summary, "a{s(uub)}")
+          dbus_signal(:SummaryUpdated, "progress:a{s(uub)}")
+        end
+
+        # @return [Boolean]
+        attr_reader :running
+
+        # @return [Integer] zero if still running
+        attr_reader :exit_code
+
+        # @return [Array<Y2390::Dasd>]
+        attr_reader :dasds
+
+        # Constructor
+        #
+        # @param initial [Array<Y2S390::FormatStatus>] initial status report from the format process
+        # @param dasds_tree [DasdsTree] see #dasds_tree
+        # @param path [DBus::ObjectPath] path in which the Job object is exported
+        # @param logger [Logger, nil]
+        def initialize(initial, dasds_tree, path, logger: nil)
+          super(path, logger: logger)
+
+          @exit_code = 0
+          @running = true
+          @dasds_tree = dasds_tree
+          @infos = {}
+          update_info(initial)
+        end
+
+        # Current status, in the format described by the D-Bus API
+        def summary
+          result = {}
+          @infos.each_value { |i| result[i.path] = i.to_dbus if i.path }
+          result
+        end
+
+        # Marks the job as finished
+        #
+        # @note A Finished signal is always emitted.
+        #
+        # @param exit_code [Integer]
+        def finish_format(exit_code)
+          @running = false
+          @exit_code = exit_code
+          Finished(exit_code)
```

Comment:
D_Bus pedantry not to confuse clients that cache properties:

```suggestion
          Finished(exit_code)
          dbus_properties_changed(JOB_INTERFACE, interfaces_and_properties.slice(JOB_INTERFACE), [])          
```

Here we use [DBus::Object#interfaces_and_properties](https://www.rubydoc.info/gems/ruby-dbus/0.19.0/DBus/Object#interfaces_and_properties-instance_method) combined with [Hash#slice](https://docs.ruby-lang.org/en/2.7.0/Hash.html#method-i-slice), neatly shortening this more obvious alternative:

```rb
dbus_properties_changed(JOB_INTERFACE, {"Running" => @running, "ExitCode" => @exit_code}, [])
```